contracts/helpers/TransferHelper.sol

```diff
@@ -68,97 +80,223 @@ contract TransferHelper is TransferHelperInterface, TokenTransferrer {
-                            recipient,
-                            item.amount
-                        );
-                    } else if (item.itemType == ConduitItemType.ERC721) {
-                        _performERC721Transfer(
-                            item.token,
-                            msg.sender,
-                            recipient,
-                            item.identifier
-                        );
-                    } else {
-                        _performERC1155Transfer(
-                            item.token,
-                            msg.sender,
-                            recipient,
-                            item.identifier,
-                            item.amount
-                        );
+        // Create a boolean that reflects whether recipient is a contract.
+        bool recipientIsContract = recipient.code.length != 0;
+
+        // Skip overflow checks: all for loops are indexed starting at zero.
+        unchecked {
+            // Iterate over each transfer.
+            for (uint256 i = 0; i < totalTransfers; ++i) {
+                // Retrieve the transfer in question.
+                TransferHelperItem calldata item = items[i];
+
+                // Perform a transfer based on the transfer's item type.
+                if (item.itemType == ConduitItemType.ERC20) {
+                    // Ensure that the identifier for an ERC20 token is 0.
+                    if (item.identifier != 0) {
+                        revert InvalidERC20Identifier();
                     }
-                }
-            }
-        }
-        // Otherwise, a conduitKey was provided.
-        else {
-            // Derive the conduit address from the deployer, conduit key
-            // and creation code hash.
-            address conduit = address(
-                uint160(
-                    uint256(
-                        keccak256(
-                            abi.encodePacked(
-                                bytes1(0xff),
-                                address(_CONDUIT_CONTROLLER),
-                                conduitKey,
-                                _CONDUIT_CREATION_CODE_HASH
+
+                    // Transfer ERC20 token.
+                    _performERC20Transfer(
+                        item.token,
+                        msg.sender,
+                        recipient,
+                        item.amount
+                    );
+                } else if (item.itemType == ConduitItemType.ERC721) {
+                    // If recipient is a contract, ensure it can receive
+                    // ERC721 tokens.
+                    if (recipientIsContract) {
+                        // Check if recipient can receive ERC721 tokens.
+                        try
+                            IERC721Receiver(recipient).onERC721Received(
+                                address(this),
+                                msg.sender,
+                                item.identifier,
```

Comment:
Disagree with the premise that we can call this from the transfer helper and get accurate results. Not a view function by interface which means we should accommodate state changes--likely would be a permissioned function. I'd advocate for either making this wholly "unsafe" transfers, or having a flag which allows for real safe transfers

Reply:
Why not just simply using safeTransferFrom that already perform these checks and call onERC721Received?

Reply:
The idea behind this optimization is that it would reduce gas costs.